docs/rfcs/0009-remove-interactors-and-controllers.md

```diff
@@ -0,0 +1,401 @@
+        ...
+    }
+}
+```
+
+Following this proposal, the above would change to something like:
+
+```kotlin
+// 1. in LibrarySiteItemView
+setOnClickListener {
+    when {
+        selected.isEmpty() -> store.dispatch(HistoryFragmentAction.AddItemForRemoval(item))
+        ...
+    }
+}
+
+// 2. reducer handles state
+private fun historyStateReducer(
+    state: HistoryFragmentState,
+    action: HistoryFragmentAction,
+): HistoryFragmentState {
+    return when (action) {
+        is HistoryFragmentAction.AddItemForRemoval(item) -> when (state.mode) {
+            HistoryFragmentState.Mode.Syncing -> state
+            else -> state.copy(
+                mode = HistoryFragmentState.Mode.Editing(state.mode.selectedItems + action.item)
+            )
+        }
+        ...
+    }
+}
+```
+
+This reduces the number of layers in the code path. Note also that business logic around whether to update based on Sync status is incorporated locally in the Reducer. In this example, the details of observing state from the store are complicated by the additional view hierarchies required by recycler views, so they have been omitted for brevity.
+
+---
+
+### Reacting to changes: state observations and side-effects in XML
+
+Generally speaking, there are two opportunities to launch side-effects in a Redux pattern:
+1. Reacting to a user-initiated action
+2. Reacting to the result of a state change
+
+The first will be covered in more detail below and is handled by middleware, where side-effects are started in reaction to some dispatched action.
+
+The second requires a mechanism of observation for a Store's state, triggering some logic or work in response to state updates. On Android these observations are often started in the UI layer in order to tie them to lifecycle events. However, UI framework components like activities, fragments, and views are notoriously difficult to test and we want to avoid a situation where these classes become too large or complex.
+
+There are already some existing mechanisms for doing this included in the lib state library:
+
+1. `Fragment.consumeFrom` will setup a Store observation that is triggered on _every_ state update in the Store. The [HistoryFragment updating its view hierarchy with all state updates](https://github.com/mozilla-mobile/firefox-android/blob/910afa889ebc5222a1b9a877837de5c55244d441/fenix/app/src/main/java/org/mozilla/fenix/library/history/HistoryFragment.kt#L196) is one example.
+2. `Flow.ifChanged` can be combined with things like `Fragment.consumeFlow` or `Store.flowScoped` for finer-grained state observations. This will allow reactions to specific state properties updates, like when the [HistorySearchDialogFragment changes the visibility of the AwesomeBar](https://github.com/mozilla-mobile/firefox-android/blob/910afa889ebc5222a1b9a877837de5c55244d441/fenix/app/src/main/java/org/mozilla/fenix/library/history/HistorySearchDialogFragment.kt#L189).
```

Comment:
I think this might need to be updated when https://github.com/mozilla-mobile/firefox-android/pull/2479 lands